### PR TITLE
Damerau–Levenshtein distance

### PR DESCRIPTION
**Pull request by ausi**
Adds support for transpositions, so that `AB => BA` adds a cost of only 1 instead of 2.

I made it configurable so that using a cost of `2` guarantees to still work the same way as the research paper.

This was more of a thought experiment for me, so feel free to close this PR if you don’t need it ☺️

### CODE REVIEW

**Comment by ausi** [comment]
I added my own implementation of the Damerau-Levenshtein distance in d31170ec7a50f37ee49a7b8281fe9a53637fd83b maybe you can use that to compare it against your automaton variant :)

**Review by Toflar — COMMENTED** [review]
Love it! I have just pushed a PR to use a Levenshtein-Automaton instead of actually calculating the distance for performance reasons (https://github.com/Toflar/state-set-index/pull/5). So we don't really need to have a distance function in this library anymore but I would still merge it because it can be very useful.

I have some TODOs in https://github.com/Toflar/state-set-index/pull/5, maybe you want to give those a try now that you know the algorithms by heart?

File: src/DamerauLevenshtein.php [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+     * ```
+     */
+    public static function distance(string $string1, string $string2, int $maxDistance = PHP_INT_MAX, int $insertionCost = 1, int $replacementCost = 1, int $deletionCost = 1, int $transpositionCost = 1): int
+    {
+        if ($string1 === $string2) {
+            return 0;
+        }
+
+        $string1Length = \strlen($string1);
+        $string2Length = \strlen($string2);
```
Comment (Toflar) [inline src/DamerauLevenshtein.php:36]: You should be using `mb_*` functions everywhere in order to support UTF-8 strings.

File: tests/DamerauLevenshteinTest.php [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    /**
+     * @dataProvider distanceProvider
+     */
+    public function testDistance(int $expected, string $a, string $b, int $maxDistance = PHP_INT_MAX): void
+    {
+        $this->assertSame($expected, DamerauLevenshtein::distance($a, $b, $maxDistance));
+        $this->assertSame($expected, DamerauLevenshtein::distance($b, $a, $maxDistance));
+    }
+
+    public static function distanceProvider(): \Generator {
```
Comment (Toflar) [inline tests/DamerauLevenshteinTest.php:20]: Should have some UTF-8 strings for tests, you can copy mine from the `LevenshteinTest`.

**Review by Toflar — CHANGES_REQUESTED** [review]

File: src/DamerauLevenshtein.php (hunk 1 above)
Comment (ausi) [reply src/DamerauLevenshtein.php]: Done in 0c7c3a5f82e31918555577df6536a011bc4f2893

**Review by Toflar — APPROVED** [review]

**Comment by Toflar** [comment]
Thank you, @ausi!

**Pull request merged**